### ISSUE TITLE
In mulitple parallel suborchestrators child activites may not happen in sequence

### ISSUE THREAD

**Issue by fashaikh** [Azure/azure-functions-durable-extension#587]
**Describe the bug**
Activities in a suborchestrator dont seem to happen in sequence when muiltiple instances of it are invoked in parallel.

**Investigative information**

- Durable Functions extension version: 1.7.1
- Function App version (1.0 or 2.0): 2.0
- Programming language used: C#

***If JavaScript***

- durable-functions npm module version:

***If deployed to Azure***

- Timeframe issue observed:
- Function App name:
- Function name(s):
- Region:
- Orchestration instance ID(s):

> If you don't want to share your Function App name or Functions names on GitHub, please be sure to provide your Invocation ID, Timestamp, and Region - we can use this to look up your Function App/Function. Provide an invocation id per Function. See the [Functions Host wiki](https://github.com/Azure/azure-webjobs-sdk-script/wiki/Sharing-Your-Function-App-name-privately) for more details.

**To Reproduce**

1. Git clone https://github.com/fashaikh/TrialResourceOrchestrator.git
2. SET Environment Variable AzureServicesAuthConnectionString=RunAs=App;AppId=<>;TenantId=<>;AppKey=<>  
3. give access to the MSI to a few subscriptions. (Modify config.json accordingly)

> While not required, providing your orchestrator's source code in anonymized form is often very helpful when investigating unexpected orchestrator behavior.

**Expected behavior**
Activity CreateResourceGroup  is called before activity StartDeployment

**Actual behavior**
Breakpoint gets hit in StartDeployment before CreateResourceGroup Activity causing no resourcegroup found errors

**Screenshots**
![image](https://user-images.githubusercontent.com/2650941/51643147-595a1080-1f20-11e9-8e18-fa20061d6283.png)

**Known workarounds**
none

**Additional context**

- Development environment : Visual Studio
- Links to source https://github.com/fashaikh/TrialResourceOrchestrator 
- Additional bindings used
- Function invocation IDs

**Comment by cgillum** [comment]
Thanks for sharing the source code. Can you point me directly to the parent orchestrator code and the sub-orchestrator code?

Not sure if this is related, but I noticed one of your orchestrator functions seem to be doing I/O [here](https://github.com/fashaikh/TrialResourceOrchestrator/blob/master/TrialResourceOrchestrator.FunctionApp/DurableManageResources.cs#L46), which breaks one of the primary rules of orchestrator functions since the results are non-deterministic. When the orchestrator function executes in a non-deterministic way, the behavior is undefined and will usually result in problems.

**Comment by fashaikh** [comment]
The parent orchestrator is here https://github.com/fashaikh/TrialResourceOrchestrator/blob/master/TrialResourceOrchestrator.FunctionApp/DurableManageResources.cs#L317 

which maintains a single instance of this context
https://github.com/fashaikh/TrialResourceOrchestrator/blob/master/TrialResourceOrchestrator.FunctionApp/DurableManageResources.cs#L39

and spawns mulitple sub orchestrators defined here 
https://github.com/fashaikh/TrialResourceOrchestrator/blob/master/TrialResourceOrchestrator.FunctionApp/DurableManageResources.cs#L65


BTW I removed the I/O code you pointed out and this issue stil reproes. 

Also, is there a recommended way of cleaning up all queued up activities/orchestration contexts when debugging. I usually end up manually calling /StopInstance which takes a few seconds, but in the meantime may have run some new code with old context state.

**Comment by YunchuWang** [comment]
Thanks for reporting this. We re-reviewed the original reproduction and don't see evidence that Durable Functions reordered the two activity calls within the same sub-orchestration.

Within `CreateATrialResource`, `CreateResourceGroup` is awaited before `StartDeployment` is scheduled. When multiple sub-orchestrations run in parallel, a breakpoint in `StartDeployment` for one instance can be hit before a breakpoint in `CreateResourceGroup` for a different instance, so breakpoint order without correlating the orchestration instance IDs doesn't establish an ordering violation.

The final repro also has behavior that can explain the observed failure independently of Durable activity ordering:

* `CreateResourceGroup` catches exceptions and returns successfully, so `StartDeployment` can run even when resource-group creation failed.
* The orchestrator still uses nondeterministic operations such as `DateTime.UtcNow`, `Guid.NewGuid()`, random ordering via `OrderBy(Guid.NewGuid())`, and a non-durable SDK delay. These violate the orchestrator code constraints and make replay behavior undefined.

Because the available repro is nondeterministic and no instance-correlated orchestration history or current minimal repro was provided, we're closing this as not planned. If this still occurs on a current Durable Functions version, please open a new issue with a minimal deterministic repro, the extension version, affected instance IDs, and the history showing `StartDeployment` scheduled before `CreateResourceGroup` completed for the same sub-orchestration instance.
